File: src/tests/uptilt_rounding.test.ts
import { expect, test } from '@jest/globals';
import { Coord, hasIllegalUptiltRounding, SlippiGame, getCoordListFromGame, toArrayBuffer, CheckResult } from '../index';
import { getUptiltCheck } from '../uptilt_rounding'
import * as fs from 'fs';
import * as path from 'path';

test('Test uptilt rounding (legal analog A)', async () => {
    const slpDir = path.join(__dirname, '../../test_data/legal/analog/traveltime/')
    const files: string[] = await fs.promises.readdir(slpDir);
    for (const filename of files) {
        var data = fs.readFileSync(path.join(slpDir, filename), null);
        let game = new SlippiGame(toArrayBuffer(data))
        expect(game).not.toBeNull()
        let gameCoordsPortThree: Coord[] = getCoordListFromGame(game, 2, true)
        let gameCoordsPortFour: Coord[] = getCoordListFromGame(game, 3, true)

        expect(hasIllegalUptiltRounding(game, 2, gameCoordsPortThree).result).toEqual(false)
        expect(hasIllegalUptiltRounding(game, 3, gameCoordsPortFour).result).toEqual(false)
    }
})

test('Test uptilt rounding (nonlegal analog)', async () => {
    const slpDir = path.join(__dirname, '../../test_data/nonlegal/analog/goomwave_uptilt_p1.slp')
    let data = fs.readFileSync(slpDir, null);
    let game = new SlippiGame(toArrayBuffer(data))
    expect(game).not.toBeNull()
    let gameCoordsPortOne: Coord[] = getCoordListFromGame(game, 0, true)

    let checkResult: CheckResult = getUptiltCheck(game, 0, gameCoordsPortOne)
    expect(checkResult.result).toEqual(true)
    expect(checkResult.violations.length).toBeGreaterThanOrEqual(1)
})

test('Test uptilt rounding (legal analog B)', async () => {
    const slpDir = path.join(__dirname, '../../test_data/legal/analog/Game_20250107T140347.slp')
    let data = fs.readFileSync(slpDir, null);
    let game = new SlippiGame(toArrayBuffer(data))
    expect(game).not.toBeNull()
    let gameCoordsPortOne: Coord[] = getCoordListFromGame(game, 0, true)

    let checkResult: CheckResult = hasIllegalUptiltRounding(game, 0, gameCoordsPortOne)
    expect(checkResult.result).toEqual(false)
    expect(checkResult.violations.length).toEqual(0)
})

test('Test uptilt rounding (legal analog C)', async () => {
    const slpDir = path.join(__dirname, '../../test_data/legal/analog/Game_20250107T142211.slp')
    let data = fs.readFileSync(slpDir, null);
    let game = new SlippiGame(toArrayBuffer(data))
    expect(game).not.toBeNull()
    let gameCoordsPortOne: Coord[] = getCoordListFromGame(game, 0, true)

    let checkResult: CheckResult = hasIllegalUptiltRounding(game, 0, gameCoordsPortOne)
    expect(checkResult.result).toEqual(false)
    expect(checkResult.violations.length).toEqual(0)
})

test('Test uptilt rounding (legal analog D)', async () => {
    const slpDir = path.join(__dirname, '../../test_data/legal/analog/Game_20250123T212056.slp')
    let data = fs.readFileSync(slpDir, null);
    let game = new SlippiGame(toArrayBuffer(data))
    expect(game).not.toBeNull()
    let gameCoordsPortOne: Coord[] = getCoordListFromGame(game, 0, true)

    let checkResult: CheckResult = hasIllegalUptiltRounding(game, 0, gameCoordsPortOne)
    expect(checkResult.result).toEqual(false)
    expect(checkResult.violations.length).toEqual(0)
})

File: libenforcer-wasm/tests/uptilt_rounding_test.rs
//! Integration tests for uptilt rounding check
//! Mirrors TypeScript tests in src/tests/uptilt_rounding.test.ts
//! Test count: 5

#[path = "common/mod.rs"]
mod common;

#[cfg(not(target_arch = "wasm32"))]
use libenforcer_wasm::{checks::uptilt_rounding, parser};
use peppi::io::slippi::de::read as read_slippi;
use std::io::Cursor;

use common::*;


#[test]
fn test_uptilt_rounding_legal_analog_a() {
    let files = read_slp_dir("legal/analog/traveltime/");

    for (filename, data) in files {
        let game = read_slippi(&mut Cursor::new(&data), None).unwrap();

        // Test port 3 (player index 2)
        if let Some(player_data) = parser::extract_player_data(&game, 2) {
            let result = uptilt_rounding::check(&player_data.main_coords);
            assert_eq!(
                result.result, false,
                "{} port 3 should pass uptilt rounding check",
                filename
            );
        }

        // Test port 4 (player index 3)
        if let Some(player_data) = parser::extract_player_data(&game, 3) {
            let result = uptilt_rounding::check(&player_data.main_coords);
            assert_eq!(
                result.result, false,
                "{} port 4 should pass uptilt rounding check",
                filename
            );
        }
    }
}

#[test]
fn test_uptilt_rounding_nonlegal_analog() {
    // TypeScript test calls getUptiltCheck() directly (not hasIllegalUptiltRounding),
    // because this replay's controller has too few rim coords for isBoxController
    let data = read_slp_file("nonlegal/analog/goomwave_uptilt_p1.slp");
    let game = read_slippi(&mut Cursor::new(&data), None).unwrap();
    let player_data = parser::extract_player_data(&game, 0).unwrap();

    let result = uptilt_rounding::get_uptilt_check(&player_data.main_coords);
    assert_eq!(
        result.result, true,
        "goomwave_uptilt_p1.slp should fail uptilt rounding check"
    );
    assert!(
        result.details.len() >= 1,
        "Should have at least 1 violation, got {}",
        result.details.len()
    );
}

#[test]
fn test_uptilt_rounding_legal_analog_b() {
    let data = read_slp_file("legal/analog/Game_20250107T140347.slp");
    let game = read_slippi(&mut Cursor::new(&data), None).unwrap();
    let player_data = parser::extract_player_data(&game, 0).unwrap();

    let result = uptilt_rounding::check(&player_data.main_coords);
    assert_eq!(
        result.result, false,
        "Game_20250107T140347.slp should pass uptilt rounding check"
    );
    assert_eq!(result.details.len(), 0, "Should have 0 violations");
}

#[test]
fn test_uptilt_rounding_legal_analog_c() {
    let data = read_slp_file("legal/analog/Game_20250107T142211.slp");
    let game = read_slippi(&mut Cursor::new(&data), None).unwrap();
    let player_data = parser::extract_player_data(&game, 0).unwrap();

    let result = uptilt_rounding::check(&player_data.main_coords);
    assert_eq!(
        result.result, false,
        "Game_20250107T142211.slp should pass uptilt rounding check"
    );
    assert_eq!(result.details.len(), 0, "Should have 0 violations");
}

#[test]
fn test_uptilt_rounding_legal_analog_d() {
    let data = read_slp_file("legal/analog/Game_20250123T212056.slp");
    let game = read_slippi(&mut Cursor::new(&data), None).unwrap();
    let player_data = parser::extract_player_data(&game, 0).unwrap();

    let result = uptilt_rounding::check(&player_data.main_coords);
    assert_eq!(
        result.result, false,
        "Game_20250123T212056.slp should pass uptilt rounding check"
    );
    assert_eq!(result.details.len(), 0, "Should have 0 violations");
}
